### PR TITLE
docs: add the TAD

### PR DESCRIPTION
**Pull request by thomas-mauran**
This adds the Technical Architecture Document to the repository as the conclusion of this project.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,107 @@
...
+We decided to only check the token in the gateway and not in each microservice to avoid having to check the token in each microservice. We could have do a zero trust architecture where each microservice would check the token but we decided to go with the centralized approach.
+
+We use a custom keycloak config stored in [linked_realm](./config/keycloak/linkedout_realm.json) which contains the roles and the clients we use. This config is then imported in the keycloak server and saves us a lot of time since we don't have to create the roles and the clients manually.
+
+### Profile Service
+
+The Profile Service is responsible for managing user profiles within the LinkedOut platform. It handles the creation, retrieval, and modification of user profiles, as well as the management of user experiences, evaluation, and availability. This service also has to contact the recommendation service whenever a new user is created to store a neo4j node with his uuid to then later recommend special job offers for this user.
```
Comment (Kuruyia) [inline README.md:65]:
```suggestion
The Profile Service is responsible for managing user profiles within the LinkedOut platform. It handles the creation, retrieval, and modification of user profiles, as well as the management of user experiences, evaluation, and availability. 

This service also has to contact the recommendation service whenever a new user is created to store a Neo4j node with their UUID to then later recommend special job offers for this user.
```

File: README.md [hunk 2]
```diff
@@ -1 +1,107 @@
...
+
+### Notification Service
+
+- Explain the role of the notification services
+
+### Recommendation Service
+
+The Recommendation Service is responsible for providing personalized job recommendations to users based on their profiles and experiences. It leverages a graph database (Neo4j) to store and manage user profiles, experiences, and job offers, and uses graph-based algorithms to generate recommendations.
+
+Like in the job service we populate the database with the job categories, job offers and jobs. To do so we use a [cypher script](./backend/recommendation/) and a neo4j plugin named APOC (Awesome Procedures on Cypher), which provides a set of procedures and functions to enhance the capabilities of Neo4j.
```
Comment (Kuruyia) [inline README.md:84]:
```suggestion
Like in the job service, we populate the database with the job categories, job offers and jobs. To do so, we use a [Cypher script](./backend/recommendation/) and a Neo4j plugin named APOC (Awesome Procedures On Cypher), which provides a set of procedures and functions to enhance the capabilities of Neo4j.
```

File: README.md [hunk 3]
```diff
@@ -1 +1,107 @@
...
+
+LinkedOut is a user-friendly platform tailored to seasonal workers on the hunt for job opportunities.
+
+This platform offers a variety of essential features, including user profile management, job listing services, intelligent job recommendations, streamlined recruitment processes, transparent review and rating systems, and multilingual support (initially in French and English).
+
+## Microservices
+
+One of the main goal of this project was to learn how to create and manage a microservice architecture.
+
+![microservices](./docs/ms.png)
```
Comment (Kuruyia) [inline README.md:13]:
```suggestion
![Architecture of the platform](./docs/ms.png)
```

File: README.md [hunk 4]
```diff
@@ -1 +1,107 @@
...
+### Notification Service
+
+- Explain the role of the notification services
+
+### Recommendation Service
+
+The Recommendation Service is responsible for providing personalized job recommendations to users based on their profiles and experiences. It leverages a graph database (Neo4j) to store and manage user profiles, experiences, and job offers, and uses graph-based algorithms to generate recommendations.
+
+Like in the job service we populate the database with the job categories, job offers and jobs. To do so we use a [cypher script](./backend/recommendation/) and a neo4j plugin named APOC (Awesome Procedures on Cypher), which provides a set of procedures and functions to enhance the capabilities of Neo4j.
+![graph db picture](./docs/assets/graph-db-1.png)
```
Comment (Kuruyia) [inline README.md:85]:
```suggestion
![Graph database visualization](./docs/assets/graph-db-1.png)
```

File: README.md [hunk 5]
```diff
@@ -1 +1,107 @@
...
+
+In the picture above you can see the graph database after initialization. In red you see each job nodes, in yellow each job category and in blue each job offer nodes.
```
Comment (Kuruyia) [inline README.md:87]:
```suggestion
In the picture above, you can see the graph database after initialization: in red there's each job node, in yellow each job category node and in blue each job offer node.
```

File: README.md [hunk 6]
```diff
@@ -1 +1,107 @@
...
+
+When a new user profile is created, the Recommendation Service stores the user's profile as a new node in the Neo4j graph database containing only it's id. Same thing when a new experience is created. We create a new node with the experience id and link it to both a user node and a job. By doing so we have a graph of users, experiences, jobs and job offers. We then use the neo4j driver to retrieve the recommendations. We use the graph database to find the shortest path between the user and the job offer. We then return a list of job offers ids. This list of job offers ids is then used to retrieve the job offers from the job service. If there is no job offers to return the job service will return all the job offers in the database.
+![graph db picture](./docs/assets/graph-db-2.png)
```
Comment (Kuruyia) [inline README.md:90]:
```suggestion
When a new user profile is created, the Recommendation Service stores the user profile as a new node in the Neo4j graph database containing only its ID. Same thing when a new experience is created: we create a new node with the experience ID and link it to both a user node and a job. 

By doing so we have a graph of users, experiences, jobs and job offers. We then use the Neo4j driver to retrieve the recommendations. We use the graph database to find the shortest path between the user and the job offer. We then return a list of job offers IDs. This list of job offers IDs is then used to retrieve the job offers from the job service. If there is no job offers to return, the job service will return all the job offers in the database.

![Graph database visualization](./docs/assets/graph-db-2.png)
```

**Review by Kuruyia — CHANGES_REQUESTED** [review]
Looks pretty good to me so far!

I can't really think of anything more to add to this than what's already planned (maybe something about the mobile app?). In the end, we could perhaps also add the "What went right?", "What went wrong?", "How can we ensure repeating what went right?" and "How can we avoid repeating what went wrong?" points, but I'm leaving that to your discretion.

**Comment by thomas-mauran** [comment]
> Looks pretty good to me so far!
> 
> I can't really think of anything more to add to this than what's already planned (maybe something about the mobile app?). In the end, we could perhaps also add the "What went right?", "What went wrong?", "How can we ensure repeating what went right?" and "How can we avoid repeating what went wrong?" points, but I'm leaving that to your discretion.

Great Idea gonna work in that way

File: docs/tad.md [hunk 7]
```diff
@@ -86,40 +86,125 @@ This service also has to contact the recommendation service to retrieve custom j
...
 
-- we wanted to try graph databases since we never tried it before
-- explain when and how we create nodes
-- explain how the recommendation works
-- screenshot of the database nodes
-
 ## Kubernetes deployment
 
-We chose to deploy our microservices architecture on a Kubernetes cluster to take advantage of its container orchestration capabilities, scalability, and resilience and to improve our understanding of this tool.
+We chose to deploy our microservices architecture on a Kubernetes cluster to take advantage of its container orchestration capabilities, scalability, and resilience and to improve our understanding of this tool. The production environment is deployed on a k3s cluster hosted on a proxmox virtual machine at Polytech. We used Kustomize to manage the configuration of the different services and to make the deployment easier.
```
Comment (Kuruyia) [inline docs/tad.md:114]:
```suggestion
We chose to deploy our microservices architecture on a Kubernetes cluster to take advantage of its container orchestration capabilities, scalability, and resilience and to improve our understanding of this tool. The production environment is deployed on a k3s cluster hosted on a Proxmox virtual machine at Polytech. We used Kustomize to manage the configuration of the different services and to make the deployment easier.
```

File: docs/tad.md [hunk 8]
```diff
@@ -86,40 +86,125 @@ This service also has to contact the recommendation service to retrieve custom j
 
 ### Message Service
 
-- Explain the role of the message services
+The Message Service is responsible for managing messages and message channels within the LinkedOut platform. It handles the creation and retrieval of messages and message channels. By doing so, it allows the user to communicate with employers and directly exchange messages with them.
```
Comment (Kuruyia) [inline docs/tad.md:89]: Describe a bit more this section: what is a message channel? how to create them?

File: docs/tad.md [hunk 9]
```diff
@@ -86,40 +86,125 @@ This service also has to contact the recommendation service to retrieve custom j
...
+
+<div align="center">
+  <img src="./assets/app/experienceEdit.png" alt="Experience Editing" width="300">
+  <br>
+  Experiences Editingx
+</div>
+
+## Post mortem and conclusion
+
+Overall, the project was a success. We learned a lot about microservices, kubernetes, keycloak, neo4j, NATS, protobuf, and many other technologies. We also learned a lot about the importance of organization in general and how to prepare a project the right way by planning it.
```
Comment (Kuruyia) [inline docs/tad.md:203]:
```suggestion
Overall, the project was a success. We learned a lot about microservices, Kubernetes, Keycloak, Neo4j, NATS, Protocol Buffers, and many other technologies. We also learned a lot about the importance of organization in general and how to prepare a project the right way by planning it.
```

File: docs/tad.md [hunk 10]
```diff
@@ -86,40 +86,125 @@ This service also has to contact the recommendation service to retrieve custom j
...
+
+Having a good view of the project before it starts is a key point. This is done by defining the Open API, by creating detailed models of the application. Doing this allows us to enlighten all the potential dark spots of the projects and to make sure if we work in a team that we have a same definition of done for each feature.
```
Comment (Kuruyia) [inline docs/tad.md:205]:
```suggestion
Having a good view of the project before it starts is a key point. This is done by defining the OpenAPI specification, by creating detailed models of the application. Doing this allows us to enlighten all the potential dark spots of the projects and to make sure if we work in a team that we have a same definition of done for each feature.
```

**Review by Kuruyia — CHANGES_REQUESTED** [review]

**Pull request merged**